torch/distributed/pipelining/stage.py

```diff
@@ -611,26 +610,66 @@ def backward_one_chunk(
                 "input_values": input_values,
             }
 
-        self.grads_input = self.backward_maybe_with_nosync(bwd_kwargs)
+        # Custom backward function
+        if self.dw_builder:
+            self.grads_input = self.backward_maybe_with_nosync(bwd_kwargs)
+            if full_backward:
+                self.dw_builder()()
```

Comment:
a note for later (maybe add a TODO)

its not always perf equivalent to run dInput() followed by dWeight() vs running full backward.  For instance, separating the two may require longer lifetimes for any intermediate tensors, and might prevent comm/computation overlap if there is TP comm inside dInput that could have overlapped with some of dWeight.  If dw_builder is supplied, we may want to change our semantics so we are allowed to ignore the 'dw_builder' and call full_backward directly when it is a full_backward op.